### PR TITLE
Add helpers for iterating containers in a pod

### PR DESCRIPTION
**Pull request by verb**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**: Iterating through all of the containers in a pod is common, and code is often duplicated to iterate through both `Containers` and `InitContainers`. This will get worse with additional container types such as [Ephemeral Containers](https://features.k8s.io/277).

This PR adds helpers based on the existing `VisitPodConfigmapNames`/`VisitPodSecretNames`. There are two flavors: one utility supporting early exit and one supporting pathspec intended for use with validation.

In addition to adding `VisitContainers()`, this PR updates many of the instances where `Containers` and `InitContainers` were iterated in immediate succession. This PR explicitly avoids updating the kubelet because of the frequency with which the kubelet treats `Containers` and `InitContainers` differently. I will follow up with a separate PR to a sig-node reviewer to update the kubelet.

**Which issue(s) this PR fixes**:
WIP #27140

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by verb** [comment]
This PR forked from #59416

/assign @smarterclayton

File: pkg/api/pod/util.go [hunk 1]
```diff
@@ -542,29 +520,21 @@ func overheadInUse(podSpec *api.PodSpec) bool {
 }
 
 // procMountInUse returns true if the pod spec is non-nil and has a SecurityContext's ProcMount field set to a non-default value
-func procMountInUse(podSpec *api.PodSpec) bool {
+func procMountInUse(podSpec *api.PodSpec) (inUse bool) {
```
Comment (smarterclayton) [inline pkg/api/pod/util.go]: I'd probably prefer to still explicitly return a variable (`var inUse bool; ...; return inUse`).  Past experience / precedent within kubernetes has been that reviewers / readers struggle with naked returns.

File: pkg/volume/util/nested_volumes.go [hunk 2]
```diff
@@ -70,16 +71,18 @@ func getNestedMountpoints(name, baseDir string, pod v1.Pod) ([]string, error) {
...
-			return nil, err
-		}
-	}
-	for _, container := range pod.Spec.Containers {
-		if err := checkContainer(&container); err != nil {
-			return nil, err
+
+	var err error
+	podutil.VisitContainers(&pod.Spec, func(c *v1.Container) bool {
+		if err = checkContainer(c); err != nil {
```
Comment (smarterclayton) [inline pkg/volume/util/nested_volumes.go]: interestingly enough this might read easier as

```
err = checkContainer(c)
return err == nil
```

I won't necessarily force that, but if it reads easier and is less likely for someone to accidentally alias `if err := ...` which is a common bug

File: plugin/pkg/admission/podpreset/admission.go [hunk 3]
```diff
@@ -183,16 +185,11 @@ func safeToApplyPodPresetsOnPod(pod *api.Pod, podPresets []*settingsv1alpha1.Pod
 	if _, err := mergeVolumes(pod.Spec.Volumes, podPresets); err != nil {
 		errs = append(errs, err)
 	}
-	for _, ctr := range pod.Spec.Containers {
-		if err := safeToApplyPodPresetsOnContainer(&ctr, podPresets); err != nil {
+	pods.VisitContainers(&pod.Spec, func(c *api.Container, _ *field.Path) {
+		if err := safeToApplyPodPresetsOnContainer(c, podPresets); err != nil {
```
Comment (smarterclayton) [inline plugin/pkg/admission/podpreset/admission.go]: Related to my earlier comment, how would a reviewer tell this case apart from the "get error out of the closure" case?  Structurally if the only difference is a single `:` I worry that it's not obvious.

File: pkg/apis/core/pods/helpers.go [hunk 4]
```diff
@@ -19,9 +19,27 @@ package pods
...
+	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/fieldpath"
 )
 
+// ContainerVisitor is called with each container spec, and returns true if visiting should continue.
+type ContainerVisitor func(container *api.Container, path *field.Path)
+
+// VisitContainers invokes the visitor function with a pointer to the container
+// spec of every container in the pod spec, include ephemeral containers.
+func VisitContainers(podSpec *api.PodSpec, visitor ContainerVisitor) {
```
Comment (smarterclayton) [inline pkg/apis/core/pods/helpers.go]: Same name, different signature, super confusing.  This should really be called `VisitContainersWithPath`

**Review by smarterclayton — CHANGES_REQUESTED** [review]
some comments

File: plugin/pkg/admission/podpreset/admission.go (hunk 3 above)
Comment (verb) [reply plugin/pkg/admission/podpreset/admission.go]: reflecting on it a bit further, an even more defensive method would be to use both:
```
var retErr error
podutil.VisitContainers(&pod.Spec, func(c *v1.Container) bool {
  if err := checkContainer(c); err != nil {
    retErr = err
    return false
  }
  return true
})
```
I don't think this is really necessary, though, so I didn't include it in my update.

File: plugin/pkg/admission/alwayspullimages/admission.go [hunk 5]
```diff
@@ -85,23 +82,17 @@ func (*AlwaysPullImages) Validate(attributes admission.Attributes, o admission.O
...
+	var allErrs []error
+	pods.VisitContainersWithPath(&pod.Spec, func(c *api.Container, p *field.Path) {
+		if c.ImagePullPolicy != api.PullAlways {
+			allErrs = append(allErrs, admission.NewForbidden(attributes,
+				field.NotSupported(p.Child("imagePullPolicy"), c.ImagePullPolicy, []string{string(api.PullAlways)}),
+			))
 		}
+	})
+	if len(allErrs) > 0 {
+		// TODO: consider using utilerrors.NewAggregate(allErrs)
```
Comment (verb) [inline plugin/pkg/admission/alwayspullimages/admission.go]: @smarterclayton I didn't want to change the behavior of this plugin without discussing it with someone, but I noticed other plugins use `utilerrors.NewAggregate(allErrs)` rather than returning the first error.

File: pkg/apis/core/validation/validation.go [hunk 6]
```diff
@@ -3434,17 +3434,13 @@ func ValidateAppArmorPodAnnotations(annotations map[string]string, spec *core.Po
 }
 
 func podSpecHasContainer(spec *core.PodSpec, containerName string) bool {
-	for _, c := range spec.InitContainers {
+	var hasContainer bool
+	podshelper.VisitContainersWithPath(spec, func(c *core.Container, _ *field.Path) {
 		if c.Name == containerName {
-			return true
+			hasContainer = true
```
Comment (smarterclayton) [inline pkg/apis/core/validation/validation.go]: Unfortunate here that we miss early exit

File: plugin/pkg/admission/alwayspullimages/admission.go (hunk 5 above)
Comment (smarterclayton) [reply plugin/pkg/admission/alwayspullimages/admission.go]: Hrm, if we do this then we have to evaluate all images on all pods, vs just the first invalid one.  However, in a successful case we have to check all of them.

I'm ok with the change.

Comment (smarterclayton) [reply plugin/pkg/admission/alwayspullimages/admission.go]: But can you open a follow up and we can get the plugin authors to discuss?

Comment (verb) [reply plugin/pkg/admission/alwayspullimages/admission.go]: Opened #79378

**Comment by tedyu** [comment]
For some use case, such as PodRequestsAndLimits() :
```
for _, container := range pod.Spec.InitContainers {
   maxResourceList(reqs, container.Resources.Requests)
   maxResourceList(limits, container.Resources.Limits)
```
The operation performed on init container is different from that performed on normal container.

@verb @smarterclayton 
Do you think it is worth adding a parameter, on top of *v1.Container, which indicates the type of container: normal, init, etc.

Here is a sketch:

https://pastebin.com/6xb0PUG1

**Comment by verb** [comment]
tl;dr I'm not sure it's common enough to warrant the added complexity, but I could be convinced.

@tedyu I'm in favor of making VisitContainer more useful because it reduces repetition when one wants to iterate through *all* (current and future) container types, but in this case there's no repeated code to remove.

So far your example seems representative, so we'd replace:
```
for _, container := range pod.Spec.Containers {
	addResourceList(reqs, container.Resources.Requests)
	addResourceList(limits, container.Resources.Limits)
}
// init containers define the minimum of any resource
for _, container := range pod.Spec.InitContainers {
	maxResourceList(reqs, container.Resources.Requests)
	maxResourceList(limits, container.Resources.Limits)
}
```

with:
```
podutil.VisitContainers(&pod.Spec, func(c *Container, t ContainerType) bool {
	switch t {
	case podutil.Container:
		addResourceList(reqs, container.Resources.Requests)
		addResourceList(limits, container.Resources.Limits)
	// init containers define the minimum of any resource
	case podutil.InitContainer:
		maxResourceList(reqs, container.Resources.Requests)
		maxResourceList(limits, container.Resources.Limits)
	}
	return true
})
```

I usually don't like this because we've introduced a Kubernetes-specific abstraction for no real reason. Any Go programmer can read the range statements, but they'd have to know about VisitContainers() specifically, and introducing closures adds room for bugs.

That being said, it can be easily inferred what VisitContainers does here and I think it might even be more readable if you're familiar with VisitContainers. It would make adding a new container type safer because we can easily reference all locations we may need to update for the new type.

For me it really hinges on whether it would see use in places that frequently iterate the container lists to different effect, like the kubelet. @yujuhong what do you think?

---
Side thought:

I can imagine a reviewer objecting to the unnecessary iteration of the other container types (in this case, ephemeral containers). I think that's an acceptable trade off for readability until analysis shows this to be the bottleneck, but that thinking led me to omit the early return from VisitContainersWithPath().

We could solve that with a bitmask, something like:

```
VisitContainers(&pod.Spec, Container|InitContainer, func(c *Container, t ContainerType) bool {
    switch t {
    case InitContainer: maxResourceList()
    ...
```

but I think it's probably premature optimization.

**Comment by tedyu** [comment]
Thanks @verb for detailed reply.

For part 1, I agree with you - that's why I didn't send out PR.

My understanding is that, if operations for two out of 3 (maybe more in the future) container types is common, the new parameter would help reduce duplicate code.

I will continue looking for proper use case(s).

**Comment by tedyu** [comment]
For part 2, I think passing bitmask allows skipping feature gate checking for new container type(s).
That seems to be a benefit.

**Comment by verb** [comment]
> It would make adding a new container type safer because we can easily reference all locations we may need to update for the new type.

Thinking about this a little more, if our goal was to consolidate container iterations for correctness, then the bitmask would be required.

**Comment by verb** [comment]
Something for the record: @msau42 better statement the argument of using VisitContainers for correctness than I did in https://github.com/kubernetes/kubernetes/pull/81838#issuecomment-524377332, I think.
